frontend/simple/views/Invite.vue

```diff
@@ -86,105 +33,109 @@
-          this.members.push({ state, contractId })
-        }
-        this.searchUser = null
-        this.userErrorMsg = ''
-      } catch (err) {
-        console.log(err)
-        this.userErrorMsg = L('Invalid User')
-      }
+  computed: {
+    isProposal () {
+      return this.memberCount() >= 3
     },
-    remove (index) {
-      this.members.splice(index, 1)
+    ...mapState([
+      'currentGroupId',
+      'loggedIn'
+    ]),
+    ...mapGetters([
+      'currentGroupState',
+      'memberCount'
+    ])
+  },
+  methods: {
+    updateInvitees (payload) {
+      this.form.errorMsg = null
+      Object.assign(this.form, payload.data)
     },
     async submit () {
       try {
-        this.errorMsg = null
-        // TODO: as members are successfully invited display in a
-        // seperate invitees grid and add them to some validation for duplicate invites
-        for (let member of this.members) {
-          // We need to have the latest mailbox attribute for the user
-          const mailbox = await backend.latestHash(member.state.attributes.mailbox)
-          const sentDate = new Date().toString()
+        this.form.errorMsg = null
+        const groupId = this.currentGroupId
+        const groupName = this.currentGroupState.groupName
+
+        for (let member of this.form.invitees) {
+          let mailbox = member.state.attributes.mailbox
+          let memberName = member.state.attributes.name
 
-          // We need to post the invite to the users' mailbox contract
-          const invite = new Events.HashableMailboxPostMessage(
-            {
-              from: this.$store.getters.currentGroupState.groupName,
-              headers: [this.$store.state.currentGroupId],
-              messageType: Events.HashableMailboxPostMessage.TypeInvite,
-              sentDate
-            },
-            mailbox
-          )
-          await backend.publishLogEntry(member.state.attributes.mailbox, invite)
+          if (this.isAlreadyInvited(member)) {
+            this.form.errorMsg = template(L('Failed to invite users: member {name} is already invited'),
+              { name: memberName })
+            return
+          }
 
-          // We need to make a record of the invitation in the group's contract
-          const latest = await backend.latestHash(this.$store.state.currentGroupId)
-          const invited = new Events.HashableGroupRecordInvitation(
-            {
-              username: member.state.attributes.name,
-              inviteHash: invite.toHash(),
-              sentDate
-            },
-            latest
-          )
-          await backend.publishLogEntry(this.$store.state.currentGroupId, invited)
+          let inviteToMailbox = await sbp('proposals/invite/mailRecord/create', mailbox, groupName, groupId)
+          let inviteToGroup = await sbp('proposals/invite/groupRecord/create', inviteToMailbox.toHash(), memberName, groupId)
```

Comment:
I'm thinking it's better to put all of the events in `shared/events.js` into their own domain, and have `proposals` be a selector path segment underneath that.

So something like `groupincome/contracts/proposals/...`, or `contracts/proposals/...`.

Designing selectors should be treated as a very slow art form, requiring a _lot_ of thought.

To design selectors correctly can take potentially hours of thought, because ultimately the entire program will be reduced to a series of selector calls, and the critical thing to understand about this is that unlike function names or method names, **these selectors will be set in stone for the rest of the application's lifespan!**

So you must consider the entire application and all of its libraries, components, and even nodejs/system calls when designing even a single selector!

The following is an example of two related thought-streams on how I personally think about selectors:

## Thought-stream 1

So for example, why did I suggestion a parent `groupincome/` domain above instead of just having it be `contracts/`?

Well because the selector is your path into the _entire digital realm!_

Literally everything will be accessible through the selector, whether it's a random third-party API, the ethereum network, your kernel settings — even the programming language itself!

So the top-most domain will be competing with all of that stuff!

If your application does "I/O" for example, you can't group your selectors under an `io` domain, because `io` is common and can be done by the operating system, a bunch of libraries, etc.

And since `contracts` can exist in a third-party library, we don't want to make that the domain either.

It may make sense to have domains refer to extremely high-level entities, for example:

- Every program has 1 program: itself, and this represents a domain within which functionality exists, so having a top-level domain like `groupincome` makes sense in order to access all group income functionality
- Every program exists within an environment, and it has exactly 1 of those too, so it makes sense to have maybe an `os` and/or and `environment` domain.
- Every program can communicate with other programs and needs a domain for doing that
- Every program can import libraries and call functionality within them, so it needs a domain for doing that
- Every program interfaces with some sort of programming language environment (e.g. the JVM, nodejs, any meta-language features the language has, etc.)

So these things make sense for top-level domains.

Note that `data`, `events`, and `proposals` are not sufficiently high-level enough to be considered domains, because `data`, `events`, and `proposals` are all things that can exist in millions of places with respect to the program.

Every `kernel` has "data" and "events". Every library have "data", and probably has "events". Every program has "data", and "events", etc.

So none of these make for good domains.

Or do they...?

## Thought-stream 2

Perhaps, precisely *because* of how common concepts like `data`, `events`,  and `contracts` are, we should give up on trying to force domains to always refer to high-level, unique concepts.

Perhaps the right approach with SBP is to instead create a blacklist of reserved top-level keywords that cannot be used as domains.

So for example, we could reserve words like `kernel`, `environment`, `os`, `libraries`, `language`, `io`, etc. for future use by SBP and not allow them to be registered.

And/or we can have a convention where domains starting with a character like an underscore are by-definition reserved domains. So `os` becomes `_os`, etc.

## Conclusion

Either way, this should be `proposals`, it should be all underneath `contracts`, I think. :)

Reply:
2 thought-streams as reaction :)

first, about our current development flow and the set-in-stone moment: while I was first working with these selectors, it felt like an unnecessarily heavy burden that because of the nature of SBP, everything that I come up with _now_ will be set in stone forever and thus have to be perfect on the first try. what I came up with to solve this tension is this: that set-in-stone moment will actually be the time when we publish the first prototype, and people actually start using it. that is when we should freeze our selectors. until then, as we create more and more domains, it is unavoidable to rethink the selector system a few times, as we learn and run into new issues. so yes, these selectors do require a lot of thought, but I would rather face this in an iterative manner, reacting to new tensions on the go, than trying to come up with the perfect solution in one sitting, trying to foresee said tensions. **please confirm or correct this :)**

and second, about the actual selector thought process: what this reminds me is the often neglected art of defining CSS selectors for styling. the question most people forget there is: is the parent I'm adding actually relevant? would this item behave/look the same it it weren't inside this parent? I think this is very important when thinking about SBP namespaces and specificity as well. could that event handling domain be reusable, even outside GroupIncome? would it work the same in another project? is it really domain-specific to GroupIncome, or could it be some distinct, reusable managing logic on its own? if it could, it should!

I understand that namespacing would be important then as well to avoid domain clashes, but then we should come up with another name for our reusable utility domains, not related to GroupIncome, so it won't be confusing when we - or others - start reusing it in other projects.

Reply:
and I almost missed the conclusion: as I see it now, yes, `proposals` should be under `groupincome` and under `contracts` as they are domain specific to GroupIncome, but I would rather keep `events` apart from this as that seems like a nice target for a reusable domain. what do you think?

Reply:
Yes, I very much agree with your thinking.

> but I would rather face this in an iterative manner, reacting to new tensions on the go, than trying to come up with the perfect solution in one sitting, trying to foresee said tensions. please confirm or correct this :)

Sure, perhaps PRs are a good place to resolve these tensions, as they allow for back and forth conversation from multiple perspectives about what the names should be. And it doesn't have to be every PR, or only in PRs. Perhaps a reviewer realizes only after two PRs that the selectors could be different, or the developer recognizes themselves a better naming system after chatting on Slack, etc.

> but I would rather keep `events` apart from this as that seems like a nice target for a reusable domain. what do you think?

I'm not so sure... I do see what you mean, that `events` can be used not just in group income, but elsewhere too. But it's also the case that even within groupincome we might have to interact with multiple different event systems from either different imported libraries or different aspects of the parent system/environment.

So, I honestly don't know. Maybe weigh the different pros/cons for different approaches?

For example, another possibility is to name it `hubudibu/events`, or `aniko/events`, if you want it to be an events library that others can use...